File: sdk-kotlin/src/main/kotlin/trade/decibel/sdk/examples/Integration.kt
/**
 * Integration example: list markets, check balance, place a trade, check balance again.
 *
 * Usage:
 *   export DECIBEL_PRIVATE_KEY="0x..."
 *   export DECIBEL_ACCOUNT_ADDRESS="0x..."
 *   export APTOS_NODE_API_KEY="..."          # optional
 *   ./gradlew run --args="integration"
 *
 * Or run this main function directly from your IDE.
 */
package trade.decibel.sdk.examples

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.delay
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// ── Inline models (mirrors DecibelSDK types) ────────────────────────────

@Serializable
data class MarketInfo(
    @SerialName("market_addr") val marketAddr: String,
    @SerialName("market_name") val marketName: String,
    @SerialName("max_leverage") val maxLeverage: Double,
    @SerialName("tick_size") val tickSize: Double,
    @SerialName("lot_size") val lotSize: Double,
    @SerialName("min_size") val minSize: Double,
)

@Serializable
data class PriceInfo(
    val market: String,
    @SerialName("mark_px") val markPx: Double,
    @SerialName("mid_px") val midPx: Double,
    @SerialName("oracle_px") val oraclePx: Double,
    @SerialName("funding_rate_bps") val fundingRateBps: Double,
)

@Serializable
data class BookLevel(val price: Double, val size: Double)

@Serializable
data class DepthInfo(
    val market: String,
    val bids: List<BookLevel>,
    val asks: List<BookLevel>,
)

@Serializable
data class AccountInfo(
    @SerialName("perp_equity_balance") val perpEquityBalance: Double,
    @SerialName("total_margin") val totalMargin: Double,
    @SerialName("unrealized_pnl") val unrealizedPnl: Double,
    @SerialName("usdc_cross_withdrawable_balance") val withdrawable: Double,
)

@Serializable
data class OrderInfo(
    @SerialName("order_id") val orderId: String,
    val market: String,
    val price: Double,
    @SerialName("remaining_size") val remainingSize: Double,
    @SerialName("is_buy") val isBuy: Boolean,
)

// ── Main ────────────────────────────────────────────────────────────────

suspend fun main() {
    val baseURL = "https://api.testnet.decibel.trade/api/v1"
    val apiKey = System.getenv("APTOS_NODE_API_KEY") ?: ""
    val privateKey = System.getenv("DECIBEL_PRIVATE_KEY") ?: ""
    val accountAddr = System.getenv("DECIBEL_ACCOUNT_ADDRESS")
        ?: "0x0000000000000000000000000000000000000000000000000000000000000001"

    val json = Json { ignoreUnknownKeys = true }
    val client = HttpClient(CIO) {
        install(ContentNegotiation) { json(json) }
    }

    fun HttpRequestBuilder.auth() {
        if (apiKey.isNotEmpty()) header("x-api-key", apiKey)
    }

    try {
        // ── 1. List all available markets ───────────────────────────────
        println("=== Available Markets ===")
        val markets: List<MarketInfo> = client.get("$baseURL/markets") { auth() }.body()
        for (m in markets) {
            println("  %-12s  max_lev: %5.0fx  tick: %-8s  lot: %s".format(
                m.marketName, m.maxLeverage, m.tickSize, m.lotSize))
        }
        println("Total markets: ${markets.size}\n")

        if (markets.isEmpty()) {
            println("No markets found. Exiting.")
            return
        }

        val market = markets.first()

        // ── 2. Fetch current price ──────────────────────────────────────
        val prices: List<PriceInfo> = client.get("$baseURL/prices/${market.marketName}") { auth() }.body()
        prices.firstOrNull()?.let { p ->
            println("=== ${market.marketName} Prices ===")
            println("  mark: ${p.markPx}  mid: ${p.midPx}  oracle: ${p.oraclePx}  funding: ${p.fundingRateBps} bps\n")
        }

        // ── 3. Fetch order book depth ───────────────────────────────────
        val depth: DepthInfo = client.get("$baseURL/depth/${market.marketName}?limit=5") { auth() }.body()
        println("=== ${market.marketName} Order Book (top 5) ===")
        println("  Bids:")
        for (b in depth.bids) println("    ${b.size} @ ${b.price}")
        println("  Asks:")
        for (a in depth.asks) println("    ${a.size} @ ${a.price}")
        println()

        // ── 4. Check balance BEFORE trade ───────────────────────────────
        val subaccount = accountAddr
        try {
            val before: AccountInfo = client.get("$baseURL/account/$subaccount") { auth() }.body()
            println("=== Balance BEFORE Trade ===")
            println("  equity:       ${before.perpEquityBalance}")
            println("  margin:       ${before.totalMargin}")
            println("  unrealised:   ${before.unrealizedPnl}")
            println("  withdrawable: ${before.withdrawable}\n")
        } catch (e: Exception) {
            println("  (Could not fetch account overview: ${e.message})\n")
        }

        // ── 5. Place a trade (conceptual) ───────────────────────────────
        println("=== Placing Order (conceptual) ===")
        println("  market: ${market.marketName}  side: BUY  price: <10% below mid>  size: ${market.minSize}  tif: GTC")
        if (privateKey.isEmpty()) {
            println("  (Skipped — set DECIBEL_PRIVATE_KEY to submit on-chain)")
        } else {
            println("  → Would build + sign + submit Aptos transaction here.")
        }
        println()

        // ── 6. Check balance AFTER trade ────────────────────────────────
        delay(500)
        try {
            val after: AccountInfo = client.get("$baseURL/account/$subaccount") { auth() }.body()
            println("=== Balance AFTER Trade ===")
            println("  equity:       ${after.perpEquityBalance}")
            println("  margin:       ${after.totalMargin}")
            println("  unrealised:   ${after.unrealizedPnl}")
            println("  withdrawable: ${after.withdrawable}\n")
        } catch (_: Exception) { }

        // ── 7. Show open orders ─────────────────────────────────────────
        try {
            val orders: List<OrderInfo> = client.get("$baseURL/open-orders/$subaccount") { auth() }.body()
            println("=== Open Orders (${orders.size}) ===")
            for (o in orders) {
                val side = if (o.isBuy) "BUY" else "SELL"
                println("  ${o.orderId} $side ${o.market} @ ${o.price} (remaining: ${o.remainingSize})")
            }
        } catch (_: Exception) { }

        println("\nDone.")
    } finally {
        client.close()
    }
}

File: sdk-swift/Examples/Integration.swift
/// Integration example: list markets, check balance, place a trade, check balance again.
///
/// This file demonstrates the read-only flow using DecibelSDK types.
/// On-chain transaction submission (placing orders) requires Aptos SDK
/// integration which is platform-specific.
///
/// Usage:
///   swift run Integration
///   (or copy into an Xcode project that depends on DecibelSDK)

import Foundation
#if canImport(DecibelSDK)
import DecibelSDK
#endif

// MARK: - Minimal HTTP helper (self-contained example)

let baseURL = "https://api.testnet.decibel.trade/api/v1"

func fetchJSON<T: Decodable>(_ path: String, as type: T.Type) async throws -> T {
    guard let url = URL(string: baseURL + path) else {
        throw URLError(.badURL)
    }
    var request = URLRequest(url: url)
    if let key = ProcessInfo.processInfo.environment["APTOS_NODE_API_KEY"] {
        request.setValue(key, forHTTPHeaderField: "x-api-key")
    }
    let (data, response) = try await URLSession.shared.data(for: request)
    if let http = response as? HTTPURLResponse, http.statusCode >= 400 {
        let body = String(data: data, encoding: .utf8) ?? ""
        throw NSError(domain: "API", code: http.statusCode,
                      userInfo: [NSLocalizedDescriptionKey: body])
    }
    return try JSONDecoder().decode(T.self, from: data)
}

// MARK: - Inline models (mirrors DecibelSDK types)

struct MarketInfo: Codable {
    let market_addr: String
    let market_name: String
    let max_leverage: Double
    let tick_size: Double
    let lot_size: Double
    let min_size: Double
}

struct PriceInfo: Codable {
    let market: String
    let mark_px: Double
    let mid_px: Double
    let oracle_px: Double
    let funding_rate_bps: Double
}

struct BookLevel: Codable {
    let price: Double
    let size: Double
}

struct DepthInfo: Codable {
    let market: String
    let bids: [BookLevel]
    let asks: [BookLevel]
}

struct AccountInfo: Codable {
    let perp_equity_balance: Double
    let total_margin: Double
    let unrealized_pnl: Double
    let usdc_cross_withdrawable_balance: Double
}

struct OrderInfo: Codable {
    let order_id: String
    let market: String
    let price: Double
    let remaining_size: Double
    let is_buy: Bool
}

// MARK: - Main

@main
struct IntegrationExample {
    static func main() async {
        let accountAddr = ProcessInfo.processInfo.environment["DECIBEL_ACCOUNT_ADDRESS"]
            ?? "0x0000000000000000000000000000000000000000000000000000000000000001"

        do {
            // ── 1. List all available markets ───────────────────────────
            print("=== Available Markets ===")
            let markets: [MarketInfo] = try await fetchJSON("/markets", as: [MarketInfo].self)
            for m in markets {
                print("  \(m.market_name)  max_lev: \(m.max_leverage)x  tick: \(m.tick_size)  lot: \(m.lot_size)")
            }
            print("Total markets: \(markets.count)\n")

            guard let market = markets.first else {
                print("No markets found. Exiting.")
                return
            }

            // ── 2. Fetch current price ──────────────────────────────────
            let prices: [PriceInfo] = try await fetchJSON("/prices/\(market.market_name)", as: [PriceInfo].self)
            if let p = prices.first {
                print("=== \(market.market_name) Prices ===")
                print("  mark: \(p.mark_px)  mid: \(p.mid_px)  oracle: \(p.oracle_px)  funding: \(p.funding_rate_bps) bps\n")
            }

            // ── 3. Fetch order book depth ───────────────────────────────
            let depth: DepthInfo = try await fetchJSON("/depth/\(market.market_name)?limit=5", as: DepthInfo.self)
            print("=== \(market.market_name) Order Book (top 5) ===")
            print("  Bids:")
            for b in depth.bids { print("    \(b.size) @ \(b.price)") }
            print("  Asks:")
            for a in depth.asks { print("    \(a.size) @ \(a.price)") }
            print()

            // ── 4. Check balance BEFORE trade ───────────────────────────
            let subaccount = accountAddr
            if let overview: AccountInfo = try? await fetchJSON("/account/\(subaccount)", as: AccountInfo.self) {
                print("=== Balance BEFORE Trade ===")
                print("  equity:       \(overview.perp_equity_balance)")
                print("  margin:       \(overview.total_margin)")
                print("  unrealised:   \(overview.unrealized_pnl)")
                print("  withdrawable: \(overview.usdc_cross_withdrawable_balance)\n")
            } else {
                print("  (Could not fetch account overview)\n")
            }

            // ── 5. Place a trade (conceptual) ───────────────────────────
            print("=== Placing Order (conceptual) ===")
            print("  market: \(market.market_name)  side: BUY  price: <10% below mid>  size: \(market.min_size)  tif: GTC")
            if ProcessInfo.processInfo.environment["DECIBEL_PRIVATE_KEY"] == nil {
                print("  (Skipped — set DECIBEL_PRIVATE_KEY to submit on-chain)")
            } else {
                print("  → Would build + sign + submit Aptos transaction here.")
            }
            print()

            // ── 6. Check balance AFTER trade ────────────────────────────
            try await Task.sleep(nanoseconds: 500_000_000)
            if let overview: AccountInfo = try? await fetchJSON("/account/\(subaccount)", as: AccountInfo.self) {
                print("=== Balance AFTER Trade ===")
                print("  equity:       \(overview.perp_equity_balance)")
                print("  margin:       \(overview.total_margin)")
                print("  unrealised:   \(overview.unrealized_pnl)")
                print("  withdrawable: \(overview.usdc_cross_withdrawable_balance)\n")
            }

            // ── 7. Show open orders ─────────────────────────────────────
            if let orders: [OrderInfo] = try? await fetchJSON("/open-orders/\(subaccount)", as: [OrderInfo].self) {
                print("=== Open Orders (\(orders.count)) ===")
                for o in orders {
                    let side = o.is_buy ? "BUY" : "SELL"
                    print("  \(o.order_id) \(side) \(o.market) @ \(o.price) (remaining: \(o.remaining_size))")
                }
            }

            print("\nDone.")

        } catch {
            print("Error: \(error)")
        }
    }
}
